doc/unicode.md

```diff
@@ -37,35 +54,83 @@ NOT_CYRILLIC : [\P{Script=Cyrillic}] ;
+as the old {@link ANTLRFileStream}, which assumed 16-bit characters, for ASCII text.
+* Through some [serious testing and performance tweaking](https://github.com/antlr/antlr4/pull/1781), the new streams perform as fast or faster than the 4.6 streams.
+
+**WARNING**. *You should avoid using both the deprecated and the new streams* in the same application because you will see 
+a nontrivial performance degradation. This speed hit is because the 
+`Lexer`'s internal code goes from a monomorphic to megamorphic
+dynamic dispatch to get characters from the input stream. Java's
+on-the-fly compiler (JIT) is unable to perform the same optimizations
+so stick with either the old or the new streams, if performance is
+a primary concern. See the [extreme debugging and spelunking](https://github.com/antlr/antlr4/pull/1781) needed to identify this issue in our timing rig.
+
+### Character Buffering, Unbuffered streams
+
+The ANTLR character streams still buffer all the input when you create
+the stream, as they have done for ~20 years. 
+
+If you need unbuffered
+access, please note that it becomes challenging to create
+parse trees. The parse tree has to point to tokens which will either
+point into a stale location in an unbuffered stream or you have to copy
+the characters out of the buffer into the token. That defeats the purpose
+of unbuffered input. See the [ANTLR 4 book](https://www.amazon.com/Definitive-ANTLR-4-Reference/dp/1934356999) "13.8 Unbuffered Character and Token Streams". Unbuffered streams are primarily
+useful for processing infinite streams *during the parse* and require that you manually buffer characters. Use `UnbufferedCharStream` and `UnbufferedTokenStream`.
+
+```java
+CharStream input = new UnbufferedCharStream(is);
CSVLexer lex = new CSVLexer(input);
// copy text out of sliding buffer and store in tokens lex.setTokenFactory(new CommonTokenFactory(true));
TokenStream tokens = new UnbufferedTokenStream<CommonToken>(lex); CSVParser parser = new CSVParser(tokens); parser.setBuildParseTree(false);
parser.file();
+```
+
+Your grammar that needs to have embedded actions that access the tokens as they are created, but before they disappear and are garbage collected. For example,
+
+```
+data : a=INT {int x = Integer.parseInt($a.text);} ;
+```
+
+From the code comments of `CommonTokenFactory`:
+
+> That `true` in `new CommonTokenFactory(true)` indicates whether `CommonToken.setText` should be called after 
+constructing tokens to explicitly set the text. This is useful for cases
+where the input stream might not be able to provide arbitrary substrings
+of text from the input after the lexer creates a token (e.g. the
+implementation of `CharStream.getText` in
+`UnbufferedCharStream` throws an
+`UnsupportedOperationException`). Explicitly setting the token text
+allows `Token.getText` to be called at any time regardless of the
+input stream implementation.
+
+*Currently, only Java and C# have these unbuffered streams implemented*.
```

Comment:
Oops, didn't mean to mislead — Java, C#, and C++ all have unbuffered streams today. (C++ just didn't need any changes to support Unicode code points.)